chanbackup/pubsub.go

```diff
@@ -155,42 +161,55 @@ func NewSubSwapper(ctx context.Context, startingChans []Single,
 
 // Start starts the chanbackup.SubSwapper.
 func (s *SubSwapper) Start() error {
-	var startErr error
-	s.started.Do(func() {
-		log.Infof("chanbackup.SubSwapper starting")
-
-		// Before we enter our main loop, we'll update the on-disk
-		// state with the latest Single state, as nodes may have new
-		// advertised addresses.
-		if err := s.updateBackupFile(); err != nil {
-			startErr = fmt.Errorf("unable to refresh backup "+
-				"file: %v", err)
-			return
-		}
+	// Ensure we only start the SubSwapper once.
+	if !s.started.CompareAndSwap(false, true) {
+		return fmt.Errorf("SubSwapper already started")
+	}
+
+	log.Infof("chanbackup.SubSwapper starting")
+
+	// Before we enter our main loop, we'll update the on-disk
+	// state with the latest Single state, as nodes may have new
+	// advertised addresses.
+	if err := s.updateBackupFile(); err != nil {
+		// Reset the running state since we failed to start.
+		s.started.Store(false)
+		return fmt.Errorf("unable to refresh backup "+
+			"file: %v", err)
+	}
 
-		s.wg.Add(1)
-		go s.backupUpdater()
-	})
+	s.wg.Add(1)
+	go s.backupUpdater()
 
-	return startErr
+	return nil
 }
 
 // Stop signals the SubSwapper to being a graceful shutdown.
 func (s *SubSwapper) Stop() error {
-	s.stopped.Do(func() {
-		log.Infof("chanbackup.SubSwapper shutting down...")
-		defer log.Debug("chanbackup.SubSwapper shutdown complete")
+	// Ensure we only stop once using atomic compare-and-swap
+	if !s.stopped.CompareAndSwap(false, true) {
+		return fmt.Errorf("SubSwapper already stopped")
+	}
+
+	log.Infof("chanbackup.SubSwapper shutting down...")
+	defer log.Debug("chanbackup.SubSwapper shutdown complete")
+
+	close(s.quit)
+	s.wg.Wait()
 
-		close(s.quit)
-		s.wg.Wait()
-	})
 	return nil
 }
 
 // ManualUpdate inserts/updates channel states into the swapper. The updates
 // are processed in another goroutine. The method waits for the updates to be
 // fully processed and the file to be updated on-disk before returning.
 func (s *SubSwapper) ManualUpdate(singles []Single) error {
+	// Make sure the SubSwapper is running.
+	if !s.started.Load() || s.stopped.Load() {
+		return fmt.Errorf("SubSwapper is not running, cannot " +
+			"perform manual update")
+	}
```

Comment:
Note that if `Stop` is called between this line and the `select` statement below, it is still safe because `Stop` will close the `quit` channel and unblock the `select`. Therefore, there is no risk of deadlock in this case.

However, if we use the `quit` channel instead of `stopped`, including here, it would make the code easier to understand and make it clearer to readers that there's no potential for deadlock.